api/directions/views.py

```diff
@@ -2156,3 +2157,73 @@ def change_owner_direction(request):
     directions = ', '.join([str(d.pk) for d in directions])
 
     return JsonResponse({"directions": directions})
+
+
+@login_required
+def directions_result_year(request):
+    request_data = json.loads(request.body)
+    is_lab = request_data.get('isLab', False)
+    is_paraclinic = request_data.get('isParaclinic', False)
+    is_doc_refferal = request_data.get('isDocReferral', False)
+    year = request_data['current_year']
+    d1 = datetime.strptime(f'01.01.{year}', '%d.%m.%Y')
+    start_date = datetime.combine(d1, dtime.min)
+    d2 = datetime.strptime(f'31.12.{year}', '%d.%m.%Y')
+    end_date = datetime.combine(d2, dtime.max)
+    card_pk = request_data.get('card_pk', -1)
+
+    if not is_lab and not is_doc_refferal and not is_paraclinic or card_pk == -1:
+        return JsonResponse({"results": []})
+
+    if is_lab:
+        lab_podr = get_lab_podr()
+        lab_podr = [i[0] for i in lab_podr]
+    else:
+        lab_podr = [-1]
+
+    card_pk = int(card_pk)
+    confirmed_directions = get_confirm_direction_patient_year(start_date, end_date, lab_podr, card_pk, is_lab, is_paraclinic, is_doc_refferal)
+    if not confirmed_directions:
+        return JsonResponse({"results": []})
+
+    directions = {}
+
+    for d in confirmed_directions:
+        if d.direction not in directions:
+            directions[d.direction] = {
+                'dir': d.direction,
+                'date': d.ch_time_confirmation,
+                'researches': [],
+            }
+
+        directions[d.direction]['researches'].append(d.research_title)
+    return JsonResponse({"results": list(directions.values())})
+
+
+@login_required
+def results_by_direction(request):
+    request_data = json.loads(request.body)
+    is_lab = request_data.get('isLab', False)
+    # is_paraclinic = request_data.get('isParaclinic', False)
+    # is_doc_refferal = request_data.get('isDocReferral', False)
+    direction = request_data.get('dir')
+
+    directions = request_data.get('directions', [])
+    if not directions and direction:
+        directions = [direction]
+    objs_result = {}
+    if is_lab:
+        direction_result = get_laboratory_results_by_directions(directions)
+
+        for r in direction_result:
+            if r.direction not in objs_result:
+                objs_result[r.direction] = {'dir': r.direction, 'date': r.date_confirm, 'researches': [],}
+
+            if r.research_title not in objs_result[r.direction]['researches']:
+                objs_result[r.direction]['researches'].append({'title': r.research_title, 'fio': r.fio, 'dateConfirm': r.date_confirm, 'fractions': []})
+
+            len_research = len(objs_result[r.direction]['researches'])
+
+            objs_result[r.direction]['researches'][len_research - 1]['fractions'].append({'title': r.fraction_title, 'value': r.value, 'units': r.units})
```

Comment:
Если надо вставить в последний элемент, то почему бы не использовать просто -1?
```suggestion
            objs_result[r.direction]['researches'][-1]['fractions'].append({'title': r.fraction_title, 'value': r.value, 'units': r.units})
```

Reply:
Вообще почему бы не использовать словари и тут, раз группировка?
Должна быть такая логика – если хотим сгруппировать – лучше брать словарь

Reply:
Можно такие данные будет и на фронте перебрать как надо (без преобразования researches в список, если он будет словарем)